kyo-kernel/shared/src/main/scala/kyo/kernel/ArrowEffect.scala

```diff
@@ -393,9 +403,67 @@ object ArrowEffect:
         handleFirstLoop(v, Context.empty)
     end handleFirst
 
-    /** Handles an arrow effect with a stateful implementation. This combines function implementation with state management, allowing the
-      * handler to maintain and modify state as it processes operations. The state is threaded through the computation and can influence how
-      * operations are processed.
+    /** Handles an arrow effect with a loop-based approach for greater flexibility.
+      *
+      * This variant provides two key advantages over basic handle:
+      *   1. It explicitly allows introducing new effects during handling via the S2 type parameter
+      *   2. It provides control flow through the Loop abstraction to continue or terminate processing
+      *
+      * This non-stateful handleLoop is ideal when you need to perform effectful operations with access to new effects during handling, but
+      * don't need to maintain state between effect occurrences.
+      *
+      * @param effectTag
+      *   Identifies which arrow effect to handle
+      * @param v
+      *   The computation requiring the function implementation
+      * @param handle
+      *   The function implementation that returns a Loop.Outcome for each iteration
+      * @return
+      *   The computation result with the function implementation provided
+      */
+    inline def handleLoop[I[_], O[_], E <: ArrowEffect[I, O], A, S, S2](
+        inline effectTag: Tag[E],
+        v: A < (E & S)
+    )(
+        inline handle: [C] => (I[C], Safepoint ?=> O[C] => A < (E & S)) => Loop.Outcome[A < (E & S), A] < S2
+    )(
+        using
+        inline _frame: Frame,
+        inline flat: Flat[A],
+        safepoint: Safepoint
+    ): A < (S & S2) =
+        @nowarn("msg=anonymous")
+        def handleLoopLoop(v: Loop.Outcome[A < (E & S), A] < S2, context: Context)(using Safepoint): A < (S & S2) =
+            v match
+                case _: KyoSuspend[?, ?, ?, ?, ?, ?] =>
+                    v.map(handleLoopLoop(_, context))
+                case continue: Loop.Continue[A < (E & S)] @unchecked =>
```

Comment:
The new implementation also has an optimization. If the user returns `Loop.continue` without suspensions, it skips the overhead of the `map` call and recurses directly.